src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt

```diff
@@ -817,82 +669,36 @@ private class MacroExpansionServiceImplInner(
     val isExpansionModeNew: Boolean
         get() = expansionMode is MacroExpansionMode.New
 
-    private fun vfsBatchFactory(): MacroExpansionVfsBatch {
-        return MacroExpansionVfsBatchImpl(dirs.projectDirName)
-    }
-
-    private fun createExpandedSearchScope(step: Int): GlobalSearchScope {
-        val expansionDirs = (0 until step).mapNotNull {
-            expansionsDirVi.findChild(it.toString())
-        }
-        val expansionScope = GlobalSearchScopes.directoriesScope(project, true, *expansionDirs.toTypedArray())
-        return GlobalSearchScope.allScope(project).uniteWith(expansionScope)
-    }
-
     private fun processUnprocessedMacros() {
         MACRO_LOG.info("processUnprocessedMacros")
-        checkStorageConsistencyOrClearMacrosDirectoryIfNeeded()
         if (!isExpansionModeNew) return
-        class ProcessUnprocessedMacrosTask : MacroExpansionTaskBase(
+        val task = MacroExpansionTask(
             project,
-            storage,
+            modificationTracker,
             pool,
-            ::vfsBatchFactory,
-            ::createExpandedSearchScope,
-            stepModificationTracker
-        ) {
-            override fun getMacrosToExpand(dumbService: DumbService): Sequence<List<Extractable>> {
-                val mode = expansionMode
-
-                val scope = when (mode.toScope()) {
-                    MacroExpansionScope.ALL -> GlobalSearchScope.allScope(project)
-                    MacroExpansionScope.WORKSPACE -> GlobalSearchScope.projectScope(project)
-                    MacroExpansionScope.NONE -> return emptySequence() // GlobalSearchScope.EMPTY_SCOPE
-                }
-
-                val calls = runReadActionInSmartMode(dumbService) {
-                    val calls = RsMacroCallIndex.getMacroCalls(project, scope)
-                        .filter { it.isTopLevelExpansion }
-                    MACRO_LOG.info("Macros to expand: ${calls.size}")
-                    calls.groupBy { it.containingFile.virtualFile }
-
-                }
-                return storage.makeExpansionTask(calls)
-            }
-
-            override val taskType: RsTask.TaskType get() = RsTask.TaskType.MACROS_UNPROCESSED
-        }
-        submitTask(ProcessUnprocessedMacrosTask())
+            lastUpdatedMacrosAt,
+            dirs.projectDirName,
+            RsTask.TaskType.MACROS_UNPROCESSED,
+        )
+        submitTask(task)
     }
 
     private fun processChangedMacros(workspaceOnly: Boolean) {
         MACRO_LOG.info("processChangedMacros")
-        checkStorageConsistencyOrClearMacrosDirectoryIfNeeded()
         if (!isExpansionModeNew) return
 
         // Fixes inplace rename when the renamed element is referenced from a macro call body
         if (isTemplateActiveInAnyEditor()) return
 
-        class ProcessModifiedMacrosTask : MacroExpansionTaskBase(
+        val taskType = if (workspaceOnly) RsTask.TaskType.MACROS_WORKSPACE else RsTask.TaskType.MACROS_FULL
```

Comment:
It looks like now there's no sense in making difference between workspace and full update